# 01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\tee3a\3D Objects\12-2022 Instacart Basket Analysis'

# 02. import the orders_products_combined

In [3]:
df_merged_large = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data for storing any manipulated data sets', 'orders_products_combined.pkl'))

In [4]:
df_merged_large.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                1.111484         196                  1          0   both  
1                1.111484       14084                  2          0   both  
2                1.111484       12427                  3          0   both  
3                1.111484       26088                  4          0   both  
4                1.111484       26405                  5          0   both

# 03. Check the shape

In [5]:
df_merged_large.shape

(32434489, 11)

# 04.  combine the orders_products_combined dataframe with your products data set. 

Import checked products data set

In [6]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data for storing any manipulated data sets', 'products_checked.csv'), index_col = False)

In [7]:
df_prods.shape

(49693, 6)

Data cleaning- unwanted column

In [8]:
df_prods_1 = df_prods.drop(columns = ['Unnamed: 0'])

Missing values

In [9]:
df_prods_1.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
df_nan =  df_prods_1[df_prods_1['product_name'].isnull() == True]

Finding duplicates

In [11]:
df_prods_clean = df_prods_1[df_prods_1['product_name'].isnull() == False]

In [12]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [13]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [14]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [15]:
df_prods_new = df_prods_clean_no_dups

In [16]:
df_prods_new.shape

(49672, 5)

In [17]:
df_prods_new.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

inner join Merging

In [18]:
df_orders_products_merged = df_prods_new.merge(df_merged_large, how = 'inner')

In [19]:
df_orders_products_merged.shape

(32404859, 15)

In [20]:
df_orders_products_merged.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  3.240486e+07  3.240486e+07   3.240486e+07  3.240486e+07  3.240486e+07   
mean   2.559866e+04  7.119612e+01   9.919792e+00  1.198023e+01  1.710745e+06   
std    1.408400e+04  3.821139e+01   6.281485e+00  4.956554e+02  9.872988e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.559470e+05   
50%    2.530200e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.711049e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565499e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  9.999900e+04  3.421083e+06   

            user_id  order_number     order_dow  order_hour_of_day  \
count  3.240486e+07  3.240486e+07  3.240486e+07       3.240486e+07   
mean   1.029372e+05  1.714230e+01  2.738867e+00       1.342515e+01   
std    5.946610e+04  1.753532e+01  2.090077e+00       4.246380e+00   
min    1.000000e+00  1.000000e+00  0.000000e+00       0.000000e+00   
25%    5.142200e+04  5.000000e+00  1.000000e+00       1.000000e+01   
50%    1.026160e+05  1.100000e+01  3.000000e+00       1.300000e+01   
75%    1.543890e+05  2.400000e+01  5.000000e+00       1.600000e+01   
max    2.062090e+05  9.900000e+01  6.000000e+00       2.300000e+01   

       days_since_prior_order  add_to_cart_order     reordered  
count            3.240486e+07       3.240486e+07  3.240486e+07  
mean             1.046388e+01       8.352547e+00  5.895873e-01  
std              8.838640e+00       7.127071e+00  4.919087e-01  
min              0.000000e+00       1.000000e+00  0.000000e+00  
25%              4.000000e+00       3.000000e+00  0.000000e+00  
50%              7.000000e+00       6.000000e+00  1.000000e+00  
75%              1.400000e+01       1.100000e+01  1.000000e+00  
max              3.000000e+01       1.450000e+02  1.000000e+00

In [21]:
df_orders_products_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   3139998      138    prior            28          6                 11   
1   1977647      138    prior            30          6                 17   
2    389851      709    prior             2          0                 21   
3    652770      764    prior             1          3                 13   
4   1813452      764    prior             3          4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  
0                3.000000                  5          0   both  
1               20.000000                  1          1   both  
2                6.000000                 20          0   both  
3                1.111484                 10          0   both  
4                9.000000                 11          1   both

# 05. Confirm the results of the merge using the merge flag.

In [22]:
df_orders_products_merged['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

The flag frequency shows in the “_merge” column that there are only entries that have a value of “both,”. The resulting dataframe (after the merge) has 32,404,859 rows, and each of those rows have information found in both input data sets. 

# 06. Export the newly created dataframe as orders_products_merged

In [23]:
df_orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data for storing any manipulated data sets', 'orders_products_merged.pkl'))